test/clonevolume_tester.py

```diff
@@ -0,0 +1,262 @@
+import mock
+import testtools
+
+import fake_3par_data as data
+import createvolume_tester as createvolume
+from hpe3parclient import exceptions
+
+
+# Variation of CreateVolumeUnitTest. Nothing specific to do here
+class CloneVolumeUnitTest(createvolume.CreateVolumeUnitTest):
+    pass
+
+
+# This exercises online-copy path
+class TestCloneDefault(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+
+
+# TODO: Rollback is needed for created volume else unit test would fail
+class TestCloneDefaultEtcdSaveFails(CloneVolumeUnitTest):
+    pass
+
+
+# Offline copy
+class TestCloneOfflineCopy(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.createVolume.assert_called()
+        mock_3parclient.copyVolume.assert_called()
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME,
+                         # Difference in size of source and cloned volume
+                         # triggers offline copy. Src volume size is 2.
+                         "size": 20}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+        mock_3parclient.getTask.return_value = {'status': data.TASK_DONE}
+
+
+# Make copyVolume operation fail
+class TestCloneOfflineCopyFails(CloneVolumeUnitTest):
+    def check_response(self, resp):
```

Comment:
1. Here i do not see the necessary assert calls to make this negative test case to pass
2.  Err: will have some message and code

Reply:
Implementation is incomplete. Will take care.